libraries/extensions/ros2-bridge/python/src/typed/serialize.rs

```diff
@@ -31,81 +34,201 @@ impl serde::Serialize for TypedValue<'_> {
+                let array: &Int16Array = self.value.as_primitive();
+                debug_assert!(array.len() == 1);
+                let number = array.value(0);
                 serializer.serialize_i16(number)
             }
             DataType::Int32 => {
-                let int_array: Int32Array = self.value.clone().into();
-                let number = int_array.value(0);
+                let array: &Int32Array = self.value.as_primitive();
+                debug_assert!(array.len() == 1);
+                let number = array.value(0);
                 serializer.serialize_i32(number)
             }
             DataType::Int64 => {
-                let int_array: Int64Array = self.value.clone().into();
-                let number = int_array.value(0);
+                let array: &Int64Array = self.value.as_primitive();
+                debug_assert!(array.len() == 1, "array was: {:#?}", array);
+                let number = array.value(0);
                 serializer.serialize_i64(number)
             }
             DataType::Float32 => {
-                let int_array: Float32Array = self.value.clone().into();
-                let number = int_array.value(0);
+                let array: &Float32Array = self.value.as_primitive();
+                debug_assert!(array.len() == 1);
+                let number = array.value(0);
                 serializer.serialize_f32(number)
             }
             DataType::Float64 => {
-                let int_array: Float64Array = self.value.clone().into();
-                let number = int_array.value(0);
+                let array: &Float64Array = self.value.as_primitive();
+                debug_assert!(array.len() == 1);
+                let number = array.value(0);
                 serializer.serialize_f64(number)
             }
             DataType::Utf8 => {
-                let int_array: StringArray = self.value.clone().into();
+                let int_array: &StringArray = self.value.as_string();
                 let string = int_array.value(0);
                 serializer.serialize_str(string)
             }
-            DataType::List(field) => {
-                let list_array: ListArray = self.value.clone().into();
-                let values = list_array.values();
-                let mut s = serializer.serialize_seq(Some(values.len()))?;
-                for value in list_array.iter() {
-                    let value = match value {
-                        Some(value) => value.to_data(),
-                        None => {
-                            return Err(serde::ser::Error::custom(
-                                "Value in ListArray is null and not yet supported".to_string(),
-                            ))
+            DataType::List(_field) => {
+                let list_array: &ListArray = self.value.as_list();
+                let mut s = serializer.serialize_seq(Some(list_array.len()))?;
+                for root in list_array.iter() {
```

Comment:
This outer list is expected to have a length of 1, right? If so, we should check that.